src/bootstrap/bootstrap.py

```diff
@@ -584,6 +584,69 @@ def fix_bin_or_dylib(self, fname):
             print("warning: failed to call patchelf:", reason)
             return
 
+    # Download stage1 if the `compiler/` directory hasn't been modified.
+    # TODO: add a way to opt-out and always rebuild.
+    # TODO: this currently only supports python3
+    # (and IMO it would be a burden to port to python2 since the APIs changed so much, especially encoding/decoding)
+    def maybe_download_stage1(self):
+        # Look for a version to compare to based on the current commit.
+        # There are a few different cases to handle.
+        # 1. This commit is a fast-forward from master: `master - * - * - HEAD`
+        # 2. This commit and master have diverged:
+        # ```
+        #   Y - * - HEAD
+        #  /
+        # X - * - master
+        # ```
+        # In this case, we should compare to `X`.
+        # 3. `master` and `HEAD` are radically different (>100 commits, or similar). This probably means that
+        # `master` does *not* correspond to the version we want to compare to, e.g. a fork.
+        # Instead, we want to compare to `rust-lang/rust:master`, which this has to share a recent merge base with.
+        # Unfortunately, we don't know which remote corresponds to `rust-lang/rust`. Add the remote ourselves to make up for it.
+
+        # Ignore this if it fails, the remote probably exists from a previous run.
+        subprocess.run(["git", "remote", "add", "rust-lang-rust", "https://github.com/rust-lang/rust.git"], stderr=subprocess.DEVNULL)
+        subprocess.check_call(["git", "fetch", "rust-lang-rust"])
+        commit = subprocess.check_output(["git", "merge-base", "HEAD", "rust-lang-rust/master"], text=True).strip()
+
+        # Next, check if there were changes to the compiler since the ancestor commit.
+        # If so, rebuild instead of using an outdated version of the code.
+        top_level = subprocess.check_output(["git", "rev-parse", "--show-toplevel"], text=True).strip()
+        compiler = "{}/compiler/".format(top_level)
+        # This is slightly more coarse than it needs to be; it rebuild on *any* changes, not just changes to code.
+        # It does however ignore new untracked files, since they couldn't possibly affect the build without first modifying Cargo.toml or adding `mod x;`.
+        # Since the only thing that could spuriously rebuilt is whitespace in Cargo.toml or something similar, this always rebuilds to be safe.
```

Comment:
You can add a `build.rs` without changing `Cargo.toml`.

Reply:
That's annoying. Thanks, I added a check.